### PR TITLE
Codegen LeakyRelu and partially codegen LeakyReluBackward

### PR DESCRIPTION
**Pull request by lsy323**
#### Update builder functions for LeakyRelu and LeakyReluBackward:
In `LeakyReluBackward::Lower(LoweringContext* loctx)`, the scalar value of the negative slope cannot be extracted in the XLA space, so the builder function for LeakyRelu needs to be changed from `xla::XlaOp BuildLeakyRelu(xla::XlaOp input, double negative_slope);` to `xla::XlaOp BuildLeakyRelu(xla::XlaOp input, xla::XlaOp negative_slope)`. Since `BuildLeakyRelu` calls `BuildLeakyReluBackward`, `BuildLeakyReluBackward` also needs to be updated.

#### Reason to partially codegen LeakyReluBackward:
There is an assertion on the scalar value in `at::Tensor XLANativeFunctions::leaky_relu_backward(const at::Tensor& grad_output, const at::Tensor& self, const at::Scalar& negative_slope, bool self_is_result)`. With full codegen, the function will be generated in XLANativeFunctions.cpp, and the assertion cannot be added in the aten space. To keep the assertion, the aten bridge is kept in `aten_xla_type.cpp`.

The assertion is from the native aten implementation: https://github.com/pytorch/pytorch/blob/91920ee6da8b8566d0e4bdb4825bfcb4c4938ad5/aten/src/ATen/native/Activation.cpp#L179

LazyIr
```
class LeakyRelu : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::leaky_relu);
  }

  LeakyRelu(const torch::lazy::Value& self, const torch::lazy::Value& negative_slope)
      : XlaNode(torch::lazy::OpKind(at::aten::leaky_relu),
              {self, negative_slope},
              [&]() { return LeakyReluOutputShape(self, negative_slope); },
              /* num_outputs */ 1,
              torch::lazy::MHash())
  {
    
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();
    
    return ss.str();
  }

  

  bool CanBeReused(const torch::lazy::Value& self, const torch::lazy::Value& negative_slope) const {
    return false;
    }

  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;

  
  

};

class LeakyReluBackward : public XlaNode {
 public:
  static torch::lazy::OpKind ClassOpKind() {
    return torch::lazy::OpKind(at::aten::leaky_relu_backward);
  }

  LeakyReluBackward(const torch::lazy::Value& grad_output, const torch::lazy::Value& self, const torch::lazy::Value& negative_slope, const bool& self_is_result)
      : XlaNode(torch::lazy::OpKind(at::aten::leaky_relu_backward),
              {grad_output, self, negative_slope},
              [&]() { return LeakyReluBackwardOutputShape(grad_output, self, negative_slope, self_is_result); },
              /* num_outputs */ 1,
              torch::lazy::MHash(self_is_result)),
        self_is_result(self_is_result)
  {
    
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << XlaNode::ToString();
    ss << ", self_is_result=" << self_is_result;
    return ss.str();
  }

  

  bool CanBeReused(const torch::lazy::Value& grad_output, const torch::lazy::Value& self, const torch::lazy::Value& negative_slope, const bool& self_is_result) const {
    return false;
    }

  torch_xla::XlaOpVector Lower(LoweringContext* loctx) const override;

  bool self_is_result;
  

};
```

NativeFunction
```
at::Tensor XLANativeFunctions::leaky_relu(const at::Tensor & self, const at::Scalar & negative_slope) {
    
    TORCH_LAZY_FN_COUNTER("xla::");
    auto common_device = torch_xla::bridge::GetXlaDevice(self);
    TORCH_INTERNAL_ASSERT(common_device);
    
    torch_xla::XLATensorPtr lazy_self = torch_xla::bridge::GetXlaTensorOrCreateForWrappedNumber(self, *common_device);
    auto node_negative_slope = torch::lazy::LazyGraphExecutor::Get()->
                        GetIrValueForScalarFromCodegen(negative_slope, *common_device);
    torch::lazy::NodePtr node = torch::lazy::ReuseNode<LeakyRelu>(lazy_self->GetIrValue(), node_negative_slope);
    if (!node) {
        
        node = torch::lazy::MakeNode<LeakyRelu>(lazy_self->GetIrValue(), node_negative_slope);
        CacheNode(node);
    }
    
    auto result = torch_xla::bridge::AtenFromXlaTensor(
            torch_xla::XLATensor::Create(std::move(node), *common_device));
    return result;
}
```

### CODE REVIEW

**Comment by lsy323** [comment]
Since LeakyRelu takes at::Scalar as input, this PR should land after https://github.com/pytorch/pytorch/pull/92066 is landed.

Context on the Scalar IR generation causing performance degradation: https://github.com/pytorch/xla/issues/4213

File: torch_xla/csrc/aten_xla_type.cpp [hunk 1]
```diff
@@ -1405,21 +1405,22 @@ std::tuple<at::Tensor, at::Tensor> XLANativeFunctions::kthvalue(
...
 }
 
-at::Tensor XLANativeFunctions::leaky_relu(const at::Tensor& self,
-                                          const at::Scalar& negative_slope) {
-  TORCH_LAZY_FN_COUNTER("xla::");
-  return bridge::AtenFromXlaTensor(tensor_methods::leaky_relu(
-      bridge::GetXlaTensor(self), negative_slope.to<double>()));
-}
-
 at::Tensor XLANativeFunctions::leaky_relu_backward(
```
Comment (vanbasten23) [inline torch_xla/csrc/aten_xla_type.cpp:1408]: If you add `leaky_relu_backward` to the `full_codegen` section in xla_native_functions.yaml, I would expect another `leaky_relu_backward` definition in pytorch/xla/torch_xla/csrc/generated/XLANativeFunctions.cpp and therefore it should fail to build due to multiple definition. Is that correct?

Comment (lsy323) [reply torch_xla/csrc/aten_xla_type.cpp:1408]: Yes, that's right. LeakyReluBackward is partially codegen'ed, so the aten bridging function won't be auto generated. We need to customized the bridging function due to the assertion. For LeakyReluBackward, only the XLA IR node is generated in LazyIr.h

Comment (wonjoolee95) [reply torch_xla/csrc/aten_xla_type.cpp:1408]: Thanks for the answer. Just to clarify and re-iterate the answer for the original question, the `leaky_relu_backward` is added to the `ir_gen` section, which only generates the IR related definitions. So the definition of `leaky_relu_backward` is still required here.

**Review by wonjoolee95 — APPROVED** [review]
Thanks! LGTM pending CI completion.

**Review by vanbasten23 — APPROVED** [review]

**Pull request merged**